front: convert operations from NGE to OSRD

Caracol3 opened the pull request:
close #7539 

Implement create, update and delete operation from nge to osrd
  - create nge operation file
  - create ngeCreateOperation function
  - create ngeUpdateOperation function
  - change the name of the file ngeOperations.ts to export.ts
  - create in export.ts ngeOperations function
  
  - [x] Create trainrun
    - [x] Create valid trainrun with valid path => create invalid trains chedule rolling stock must be entered in OSRD

  - [x] Update trainrun
    - [x] Update trainrun path  => update trains chedule path

  - [x] Delete trainrun
    - [x] Delete trainrun => delete train schedule

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/types.ts:
```diff
@@ -65,6 +65,30 @@ export type TimeLock = {
   timeFormatter: null;
 };
 
+export type TrainrunSectionOperation = {
```
I think we can just use the existing `TrainrunSection` type here instead of re-defining a new one?

Uriel-Sautron replied:
We thought `TrainrunSection` was used for typed TrainrunSectionDTO.
We can therefore add keys to `TrainrunSection` ?

emersion replied:
Eh, indeed, there are differences between [the DTO type](https://github.com/SchweizerischeBundesbahnen/netzgrafik-editor-frontend/blob/b83f0ec2af0c4658632d03ac4497c6a38807cb75/src/app/data-structures/business.data.structures.ts#L113) and [the model type](https://github.com/SchweizerischeBundesbahnen/netzgrafik-editor-frontend/blob/b83f0ec2af0c4658632d03ac4497c6a38807cb75/src/app/models/trainrunsection.model.ts#L19)...

Which additional fields do we need that are missing from the DTO? If we need these additional fields, I agree it's better to keep the types separate to not add extraneous fields which won't be taken into account by NGE to the DTO.

Uriel-Sautron replied:
We need `targetNode`, `sourceNode`  and  `trainrun`

louisgreiner replied:
Since the `operations` need more detail, we could stick to the model type, and create a new DTO type for `netzgrafikDto`

emersion replied:
Hm. How about we do this for now?

```ts
export type OperationTrainrunSection = TrainrunSection & {
  // extra props we need here
};
```

Nit: IMHO `OperationTrainrunSection` reads a _tad_ better than `TrainrunSectionOperation`: the latter makes it sound like the whole thing is an operation.

Uriel-Sautron replied:
Done

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/ngeOperations.ts:
Suggestion: maybe this file could be renamed to `export.ts` for consistency with the other side (`import.ts`)?

Uriel-Sautron replied:
Done

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/NGE.tsx:
```diff
@@ -0,0 +1,117 @@
+      frame.contentDocument!.body.appendChild(ngeRoot);
+
+      // listens to create, update and delete operations
+      ngeRoot.addEventListener('operation', async (event: Event) => {
```
Maybe we could introduce a function which takes an NGE operation and handles it inside `ngeOperations.ts`? Basically extract this callback and move it to that file, since it's specific to the NGE → OSRD conversion.

Uriel-Sautron replied:
Done

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/NGE.tsx:
```diff
@@ -0,0 +1,117 @@
+      ngeRoot.addEventListener('operation', async (event: Event) => {
+        console.log('Operation received', (event as CustomEvent).detail);
+
+        if ((event as CustomEvent).detail.type === 'create') {
```
This could probably use a `switch` to handle the different cases.

Uriel-Sautron replied:
Done

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/ngeOperations.ts:
```diff
@@ -0,0 +1,54 @@
+import type { Distribution, TrainScheduleBase } from 'common/api/generatedEditoastApi';
```
`osrdEditoastApi` should be used instead of `generatedEditoastApi`.

emersion replied:
Ref https://github.com/OpenRailAssociation/osrd/pull/8250

Uriel-Sautron replied:
Done

Uriel-Sautron committed:
front: convert operations from NGE to OSRD
 - change the name of the file ngeOperations.ts to export.ts
 - create in export.ts ngeOperations function

 - add secondary code in playload

  - add onOperation in props NGE

  - rebase dev

  - refacto operation payloads

  - add map to prevent trainschedule update from another scenario

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/NGE.tsx:
```diff
@@ -22,29 +33,33 @@ const frameSrc = `
+
+  const [ngeRootElement, setNgeRootElement] = useState<NGEElement | null>(null);
+
+  const { getTimetableID } = useOsrdConfSelectors();
+  const timeTableId = useSelector(getTimetableID);
```
Up until now, I've designed the `NGE` component to be completely self-contained. That is, it's not tied to anything OSRD-specific. The goal is to have a clear separation of concerns: keep the OSRD-specific logic outside of the component, and have the component only define an API to manage NGE without anything else. That way the low-level NGE details (iframe management, typing etc) are isolated away from the rest of the code.

To retain this design, we'd need to remove this and instead expose a `onOperation` prop that callers can set to a callback taking an `Operation` argument.

Do you think it's better to keep this design, or do you think there are good reason to diverge from it?

Uriel-Sautron replied:
No you're right to want to separate NGE and OSRD.

emersion replied:
Here's what I had in mind:

```ts
// types.ts
export type NGEEvent = NGECreateEvent | NGEDeleteEvent | NGEUpdateEvent;

// NGE.tsx
type NGEProps = {
  dto?: NetzgrafikDto;
  onOperation?: (event: NGEEvent);
};
// ...
ngeRoot.addEventListener('operation', async (event: Event) => {
  const customEvent = event as CustomEvent;
  const op = customEvent.detail as NGEEvent;
  if (onOperation) onOperation(op);
});

// ScenarioV2.tsx
const handleNGEOperation = (op: NGEEvent) =>
  ngeOperations(op, dispatch, timeTableId);
// ...
<NGE dto={ngeDto} onOperation={handlNGEOperation} />
```

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/export.ts:
```diff
@@ -0,0 +1,105 @@
+import type { NGEDeleteEvent, NGEEvent, NGEUpdateEvent, TrainrunSectionOperation } from './types';
+
+const defautlPayloads = {
+  constraint_distribution: 'STANDARD' as Distribution,
```
Nit: I believe `as Distribution` is unnecessary here.

Uriel-Sautron replied:
We need to specify the type here, because when the payload is created typescript assigns the string type to the constraint_distribution

clarani replied:
You can type DEFAULT_PAYLOAD like this to remove the Distribution ;)
```
const DEFAULT_PAYLOAD: Pick<
  TrainScheduleBase,
  'constraint_distribution' | 'rolling_stock_name' | 'start_time'
> = {
  constraint_distribution: 'STANDARD',
  rolling_stock_name: '',
  start_time: '2024-07-15T08:00:00+02:00',
};
```

emersion replied:
Oh right good point!

Uriel-Sautron replied:
Done

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/export.ts:
```diff
@@ -0,0 +1,105 @@
+
+import type { NGEDeleteEvent, NGEEvent, NGEUpdateEvent, TrainrunSectionOperation } from './types';
+
+const defautlPayloads = {
```
- Typo: "default"
- Nit: should this be singular? There is an "s" at the end but this object represents a single payload
- I wonder if it should be UPPER_CASE instead? I'm not sure what the style rules are but I was under the impression that constants needed to be that way.

Uriel-Sautron replied:
Done, you're right about UPPER_CASE

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/export.ts:
```diff
@@ -0,0 +1,105 @@
+const defautlPayloads = {
+  constraint_distribution: 'STANDARD' as Distribution,
+  rolling_stock_name: '',
+  start_time: '2024-07-15T08:00:00+02:00',
```
In the future, we should grab this from the first NGE train run section. However, this is out-of-scope for this PR. Maybe we should add a TODO to make it clear the hardcoded value is temporary?

Uriel-Sautron replied:
Done

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/export.ts:
```diff
@@ -0,0 +1,105 @@
+    case 'create': {
+      dispatch(
+        osrdEditoastApi.endpoints.postV2TimetableByIdTrainSchedule.initiate({
+          id: timeTableId!,
```
Nit: `!` should be unnecessary here because `timeTableId` isn't nullable

Uriel-Sautron replied:
Done

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/export.ts:
```diff
@@ -0,0 +1,105 @@
+const ngeOperations = async (event: NGEEvent, dispatch: AppDispatch, timeTableId: number) => {
+  switch (event.type) {
+    case 'create': {
+      dispatch(
```
We should probably `await` here, so that the function doesn't return before we're done sending the HTTP request?

Uriel-Sautron replied:
Done

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/export.ts:
```diff
@@ -0,0 +1,105 @@
+    case 'delete': {
+      const { trainrun } = event as NGEDeleteEvent;
+      await dispatch(
+        osrdEditoastApi.endpoints.deleteV2TrainSchedule.initiate({ body: { ids: [trainrun.id] } })
```
Here we're assuming that `trainrun.id` is the OSRD train schedule ID. This works fine if the trainrun was imported from OSRD, but breaks when the trainrun was created inside NGE. Trainruns created inside NGE get an auto-incremented ID generated by NGE which don't match OSRD IDs.

(Same applies to `update`.)

I don't think we can easily fix this here, but it's potentially dangerous since we may delete completely unrelated trainruns (what NGE thinks trainrun 42 is may be a completely unrelated trainrun for OSRD).

(My plan to fix this completely would be to add a map of NGE <-> OSRD IDs.)

emersion replied:
An, ahem, quick'n'dirty way to ~~address~~ hack this would be to use a different type to host OSRD IDs, that way we can tell them apart from NGE IDs and bail out if we encounter one. Typically convert the OSRD ID to a string at import time, check if the ID is a string here, and convert it back to a number before passing it to the backend.

Please don't credit me for this if you end up implementing it.

louisgreiner replied:
resolved by creating a map that gets the train schedule instead

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/export.ts:
```diff
@@ -0,0 +1,105 @@
+  const payloads: TrainScheduleBase[] = [
+    {
+      ...defautlPayloads,
+      path: [
+        {
+          trigram: sourceTrigram,
+          secondary_code: sourceSecondaryCode,
+          id: `${sourceNode.id}`,
+          deleted: false,
+        },
+        {
+          trigram: targetTrigram,
+          secondary_code: targetSecondaryCode,
+          id: `${targetNode.id}`,
+          deleted: false,
+        },
+      ],
```
In order to not duplicate the logic to create the `path`, could we call the logic contained in `ngeUpdateOperation` from `ngeCreateOperation`? Basically the idea would be:

```ts
const payloads: TrainScheduleBase[] = [
  ngeUpdateOperation([trainrunSection]),
];
```

Uriel-Sautron replied:
Done

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/export.ts:
```diff
@@ -0,0 +1,105 @@
+  start_time: '2024-07-15T08:00:00+02:00',
+};
+
+const ngeCreateOperation = (trainrunSection: TrainrunSectionOperation) => {
```
Naming nit: `ngeCreateOperation` makes it sound like this is a variable storing an operation, but it's actually a function. Maybe `handleCreateOperation` would be a bit better? (Same applies to other functions.)

Uriel-Sautron replied:
Done

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/export.ts:
```diff
@@ -0,0 +1,105 @@
+};
+
+const ngeUpdateOperation = (trainrunSections: TrainrunSectionOperation[]) => {
+  const path = trainrunSections.flatMap((section, index) => {
```
This loop assumes that the trainrun sections have a specific format: [A B] [B C] [C D]. However this is not something guaranteed by NGE, we could get something like [B C] [A B] or even [A B] [C D].

I don't think this is necessarily something that we need to handle in this PR, but could we add a TODO about it?

louisgreiner replied:
Since non-continous trainruns (I believe you had that in mind) exist in NGE, we need to prevent that. However it is true that it is not our priority (but soon to be)

emersion replied:
[A B] [C D] is a case of discontinuous trainruns (a single train run made up of two separate paths).

[B C] [A B] is a case of a continuous train run, but with the sections in an arbitrary order.

emersion reviewed (commented):
This is better :) here are a few more comments. Let me know if something doesn't make sense!

emersion commented:
Also, this needs a rebase now that NGE import has been merged.

clarani commented on front/src/applications/operationalStudies/components/MacroEditor/export.ts:
```diff
@@ -0,0 +1,87 @@
+    }
+    case 'update': {
+      const { trainrunSections } = event as NGEUpdateEvent;
+      const [firstTrainrunSection] = trainrunSections;
```
Maybe we should add a test here to check that `trainrunSections.length > 0`, otherwise the app will panick l76
```suggestion
      const [firstTrainrunSection] = trainrunSections;
```

emersion replied:
I believe this is guaranteed by NGE. Wouldn't hurt to add a `console.assert()` or throw an error out of caution though.

clarani commented on front/src/applications/operationalStudies/components/MacroEditor/export.ts:
```diff
@@ -0,0 +1,87 @@
+          id: timeTableId,
+          body: handleCreateOperation(event.trainrunSection),
+        })
+      ).unwrap();
```
nit: what do you think of adding a const to create the payload here ? You could then remove the function `handleCreateOperation`
```suggestion
      const trainSchedulePayload = createPayloadOperation([event.trainrunSection]);
      await dispatch(
        osrdEditoastApi.endpoints.postV2TimetableByIdTrainSchedule.initiate({
          id: timeTableId,
          body: [trainSchedulePayload],
        })
      ).unwrap();
```

clarani commented on front/src/applications/operationalStudies/components/MacroEditor/export.ts:
```diff
@@ -0,0 +1,87 @@
+  start_time: '2024-07-15T08:00:00+02:00',
+};
+
+const createPayloadOperation = (trainrunSections: OperationTrainrunSection[]) => {
```
`createTrainSchedulePayload` would be a better name I think
```suggestion
const createTrainSchedulePayload = (trainrunSections: OperationTrainrunSection[]) => {
```

Uriel-Sautron replied:
Done

clarani commented on front/src/applications/operationalStudies/components/MacroEditor/export.ts:
```diff
@@ -0,0 +1,87 @@
+const createPayloadOperation = (trainrunSections: OperationTrainrunSection[]) => {
+  // TODO: check that the trainrunSections format is still compatible
+  const path = trainrunSections.flatMap((section, index) => {
+    const [sourceTrigram, sourceSecondaryCode] = section.sourceNode.betriebspunktName.split('/');
+    const [targetTrigram, targetSecondaryCode] = section.targetNode.betriebspunktName.split('/');
+    const sourceNode = {
+      trigram: sourceTrigram,
+      secondary_code: sourceSecondaryCode,
+      id: `${section.sourceNode.id}-${index}`,
+    };
+    const targetNode = {
+      trigram: targetTrigram,
+      secondary_code: targetSecondaryCode,
+      id: `${section.targetNode.id}-${index}`,
+    };
```
nit: you can create a small function to create a pathItem from a node:
```
const createPathItemFromNode = (
  { betriebspunktName: nodeName, id: nodeId }: Node,
  index: number
) => {
  const [trigram, secondaryCode] = nodeName.split('/');
  return {
    trigram: trigram,
    secondary_code: secondaryCode,
    id: `${nodeId}-${index}`,
  };
};
```
and then 
```suggestion
     const sourceNode = createPathItemFromNode(section.sourceNode, index);
    const targetNode = createPathItemFromNode(section.targetNode, index);
```

Uriel-Sautron replied:
I had already planned this refacto, but I'm waiting for this [PR](https://github.com/OpenRailAssociation/osrd/pull/8343) to be merged as there will be changes to make on it.

clarani commented on front/src/applications/operationalStudies/components/MacroEditor/export.ts:
```diff
@@ -0,0 +1,87 @@
+      secondary_code: targetSecondaryCode,
+      id: `${section.targetNode.id}-${index}`,
+    };
+    if (index === trainrunSections.length - 1) {
+      return [sourceNode, targetNode];
+    }
```
You can compute targetNode only in the if
```suggestion
    if (index === trainrunSections.length - 1) {
      const targetNode = ...
      return [sourceNode, targetNode];
    }
```

Uriel-Sautron replied:
Done

clarani commented on front/src/applications/operationalStudies/components/MacroEditor/export.ts:
```diff
@@ -0,0 +1,87 @@
+    }
+    return [sourceNode];
+  });
+  const [firstTrainrunSection] = trainrunSections;
```
I think this is more readable with `[0]`. Should we add a test to check that `trainrunSections.length > 0` ?
```suggestion
  const firstTrainrunSection = trainrunSections[0];
```

emersion replied:
Ref https://github.com/OpenRailAssociation/osrd/pull/8220/files#r1707093806 for the discussion about the length check.

Agreed that `[0]` is more readable!

Uriel-Sautron replied:
Done

clarani commented on front/src/applications/operationalStudies/components/MacroEditor/export.ts:
```diff
@@ -0,0 +1,87 @@
+    return [sourceNode];
+  });
+  const [firstTrainrunSection] = trainrunSections;
+  const payloads: TrainScheduleBase = {
```
You can also directly return here and add the TrainScheduleBase type in the function signature
```suggestion
  const payload: TrainScheduleBase = {
```

Uriel-Sautron replied:
Done

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/NGE.tsx:
```diff
@@ -42,9 +48,15 @@ const NGE = ({ dto }: { dto?: NetzgrafikDto }) => {
-      // ngeRoot.addEventListener('operation', (event: Event) => {
-      //   console.log('Operation received', (event as CustomEvent).detail);
-      // });
+      ngeRoot.addEventListener('operation', async (event: Event) => {
```
Nit: this doesn't need to be `async` now

Uriel-Sautron replied:
Done

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/NGE.tsx:
```diff
@@ -42,9 +48,15 @@ const NGE = ({ dto }: { dto?: NetzgrafikDto }) => {
+        if (onOperation) onOperation(op);
+      });
+
+      if (dto) {
+        ngeRoot.netzgrafikDto = dto;
+      }
```
This should be removed: it was part of the old logic, but is now handled by the second `useEffect` below.

(Probably remnant from a rebase.)

Uriel-Sautron replied:
Done

clarani commented on front/src/applications/operationalStudies/views/v2/ScenarioV2.tsx:
```diff
@@ -165,6 +169,8 @@ const ScenarioV2 = () => {
     setCollapsedTimetable(isMacroMode);
   };
 
+  const handleNGEOperation = (op: NGEEvent) => ngeOperations(op, dispatch, timetableId!);
```
I think this function should be moved in `NGE.tsx`. You can give a new props `timetableId` to `NGE`, and get the dispatch with `useAppDispatch` directly in `NGE.tsx`

emersion replied:
The motivation for this to be out of `NGE.tsx` is explained here: https://github.com/OpenRailAssociation/osrd/pull/8220#discussion_r1701793216

clarani replied:
Hm I understand. I didn't want to add this function in Scenario because this component is already huge and handles a lot of high level things. But I also understand your point.
We can keep it here for the moment and move it to an NGE wrapper if more props need to be created for NGE in the future.

emersion replied:
Right! We've tried to keep the `ScenarioV2` changes as small as possible.

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/export.ts:
```diff
@@ -0,0 +1,87 @@
+const handleUpdateOperation = (trainrunSections: OperationTrainrunSection[]) =>
+  createPayloadOperation(trainrunSections);
+
+const ngeOperations = async (event: NGEEvent, dispatch: AppDispatch, timeTableId: number) => {
```
Can we also rename this to `handleOperation` for consistency?

Uriel-Sautron replied:
Done

emersion commented on front/src/applications/operationalStudies/views/v2/ScenarioV2.tsx:
```diff
@@ -8,7 +8,13 @@ import { useSelector } from 'react-redux';
 import { useParams } from 'react-router-dom';
 
 import BreadCrumbs from 'applications/operationalStudies/components/BreadCrumbs';
+import ngeOperations from 'applications/operationalStudies/components/MacroEditor/export';
```
Ditto!

Uriel-Sautron replied:
Done

emersion reviewed (commented):
Some very minor comments, apart from these and https://github.com/OpenRailAssociation/osrd/pull/8220/files#r1705446045, looks good!

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/export.ts:
```diff
@@ -0,0 +1,87 @@
+  return payloads;
+};
+
+const handleCreateOperation = (trainrunSection: OperationTrainrunSection) => [
+  createPayloadOperation([trainrunSection]),
+];
+
+const handleUpdateOperation = (trainrunSections: OperationTrainrunSection[]) =>
+  createPayloadOperation(trainrunSections);
```
Another nit: TBH, sounds like these are simple enough to just be inlined in the caller, WDYT?

Uriel-Sautron replied:
This will be the case in the next commit when the NGE upgrade PR is merged

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/export.ts:
```diff
@@ -0,0 +1,87 @@
+    return [sourceNode];
+  });
+  const [firstTrainrunSection] = trainrunSections;
+  const payloads: TrainScheduleBase = {
```
Okay, last nit: this is a single payload so should be singular not plural :P

emersion replied:
Damn it, @clarani ended up being faster than me: https://github.com/OpenRailAssociation/osrd/pull/8220/files#r1707204029

clarani commented on front/src/applications/operationalStudies/components/MacroEditor/export.ts:
`export` is not really understandable at first sight. Do you think `convertNgeOperationToOsrdOperation` is a better name ? or `ngeOperationToOsrdOperation` ?

emersion replied:
`export` is for symmetry with `import`.

clarani replied:
`export` and `import` are both not really understandable at first sight.
I would prefer to have `ngeOperationToOsrdOperation` and `osrdOperationToNgeOperation`, or more simply `ngeToOsrd` and `osrdToNge`, what do you think ?

emersion replied:
The "operation" bit is wording specific to the NGE → OSRD side. The "operation" data structures represent NGE changes/updates. So I don't think it makes sense to include in the name.

If you don't like import/export, I think ngeToOsrd/osrdToNge are great names!

clarani commented on front/src/applications/operationalStudies/components/MacroEditor/types.ts:
```diff
@@ -153,3 +159,20 @@ export type NetzgrafikDto = {
+
+export type NGEDeleteEvent = {
+  type: 'delete';
+  trainrun: Trainrun;
```
nit: you can only give the trainrunId to delete
```suggestion
  trainrunId: number;
```

emersion replied:
We can't change this without changing NGE. These data structures describe values coming from NGE.

The full trainrun isn't useful to us, but might be to other NGE users.

clarani replied:
OK got it, it's fine for me :+1:

clarani reviewed (commented):
Thanks for you PR :) I left some comments ;)

clarani commented:
I have this bug, I can't edit the name of a created train on NGE
[Capture vidéo du 2024-08-07 17-22-07.webm](https://github.com/user-attachments/assets/e3397f94-7073-4a10-9c28-c3a46d47b486)

emersion commented:
> I have this bug, I can't edit the name of a created train on NGE

I don't think we get an event from NGE for this action yet. This will be fixed once we upgrade NGE to include https://github.com/SchweizerischeBundesbahnen/netzgrafik-editor-frontend/pull/213.

@louisgreiner can you confirm?

louisgreiner commented:
Other events than:
- create a trainrun
- update the PATH of a trainrun
- delete a trainrun 

are not implemented yet as mentionned above, but will be at the end of next week, @clarani

Uriel-Sautron commented:
Thank you for your review @clarani.
We don't have a name edition event yet.

Commit:
front: handle empty rollingstock.name when editing a train schedule

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/export.ts:
```diff
@@ -0,0 +1,114 @@
+          body: { ids: [trainrunIdToDelete] },
+        })
+      ).unwrap();
+      CREATED_TRAINRUN.delete(trainrun.id);
```
Note to self: even if the map keys are NGE-generated IDs, this should be fine to do unconditionally even with IDs coming from OSRD. Indeed, NGE will never generate an ID already in use by a trainrun imported from OSRD.

louisgreiner replied:
We could indeed populate the map as the import (`osrdToNge`) is called, and then use this table only to communicate with train schedule operations

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/export.ts:
```diff
@@ -0,0 +1,114 @@
+
+import type { NetzgrafikDto, NGETrainrunEvent, TrainrunSection, Node, Trainrun } from './types';
+
+const CREATED_TRAINRUN = new Map<number, number>();
```
Tiny nit: since this is a map that we mutate at runtime (adding/removing elements as NGE operations are handled), it's not really a constant value, so maybe it's unfair to put it in UPPER_CASE?

Uriel-Sautron replied:
Done

clarani commented on front/src/applications/operationalStudies/hooks/useSetupItineraryForTrainUpdate.ts:
```diff
@@ -50,6 +50,17 @@ const useSetupItineraryForTrainUpdate = (
+      if (!trainSchedule.rolling_stock_name) {
+        adjustConfWithTrainToModifyV2(
+          trainSchedule,
+          [],
```
Why do you reset the pathSteps here ?

clarani replied:
If you want to reset the pathSteps, you need to give `[null, null]`

emersion reviewed (approved):
With the remaining comments fixed, this LGTM!

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/helpers/adjustConfWithTrainToModify.ts:
```diff
@@ -9,8 +9,8 @@ import { msToKmh } from 'utils/physics';
   trainSchedule: TrainScheduleResult,
-  pathSteps: PathStep[],
-  rollingStockId: number,
+  pathSteps: (PathStep | null)[],
```
~~You don't need to change the type here I think~~
```suggestion
  pathSteps: PathStep[],
```

clarani replied:
My bad, no you need to keep this sorry ! You can resolve this comment

emersion commented on front/src/applications/operationalStudies/hooks/useSetupItineraryForTrainUpdate.ts:
```diff
@@ -50,6 +50,17 @@ const useSetupItineraryForTrainUpdate = (
         id: trainIdToEdit,
       }).unwrap();
+      if (!trainSchedule.rolling_stock_name) {
+        adjustConfWithTrainToModifyV2(
```
Note: this fixes the edit form not getting populated with the train schedule details in case there is no rolling stock, however there are more error cases below where we should populate the form but don't do it, for instance when pathfinding fails. We should really fix that but it's orthogonal to this PR.

clarani reviewed (approved):
LGTM :white_check_mark:
